Blog for stable revision tags (#9780)
istio/istio.io: content/en/blog/2021/revision-tags/index.md

---
title: "Istio control plane canary with revisions and tags"

> description: Learn how to take more control over your Istio control-plane upgrades.

Reviewer: I do prefer Craig's suggested change. Istio control plane canary doesn't make sense to me.

publishdate: 2021-05-20
attribution: "Christian Posta (Solo.io), Lin Sun (Solo.io), Sam Naser (Google)"
keywords: [upgrades,revisions,operations,canary]
---

> The Istio community [releases new versions of Istio](/about/supported-releases/) every quarter and sometimes between the normal release cadence for bug fixes [or CVE patches](/about/security-vulnerabilities/). The operator of an Istio service mesh needs to upgrade the control plane and data plane components during the life of the service mesh. You must take care when upgrading a service mesh as a mistake could take down the entire mesh and affect your business traffic. Istio has a couple mechanisms to make it safe to perform upgrades in a controlled manner, and in Istio 1.10 we further improve this operational experience.

Reviewer 2: ```suggestion
Like all security software, your service mesh should be kept up-to-date. The Istio community [releases new versions every quarter](/docs/releases/supported-releases/), with regular patch releases for bug fixes [and security vulnerabilities](/docs/releases/security-vulnerabilities/). The operator of a service mesh will need to upgrade the control plane and data plane components many times. You must take care when upgrading, as a mistake could affect your business traffic. Istio has many mechanisms to make it safe to perform upgrades in a controlled manner, and in Istio 1.10 we further improve this operational experience.
```

Reviewer: Changes here should fix the lint issue with links.

## Background

> In Istio 1.6, [we added basic support for upgrading the service mesh](/news/releases/1.6.x/announcing-1.6/change-notes/) following a [canary pattern using revisions](/blog/2020/multiple-control-planes/). Using this approach, you can run multiple control planes side-by-side without impacting an existing deployment and slowly migrate workloads from an old control plane to the new control plane.

Reviewer 2: ```suggestion
[In Istio 1.6](/news/releases/1.6.x/announcing-1.6/change-notes/), we added [basic support for upgrading the service mesh following a canary pattern, using revisions](/blog/2020/multiple-control-planes/). Using this approach, you can run multiple control planes side-by-side without impacting an existing deployment and slowly migrate workloads from the old control plane to the new.
```

> To support this revision-based upgrade, Istio introduced the `istio.io/rev` label for namespaces which indicates to which control plane the workload data-plane proxies in that namespace should connect. For example, a label of `istio.io/rev=1-9-3` indicates the control plane revision `1-9-3` should inject the data plane for workloads in that namespace.
> 
> If you wanted to upgrade the data-plane proxies for a particular namespace, you would update the `istio.io/rev` label to point to a new version such as `istio.io/rev=1-10-0`. Across many namespaces, updating labels one at a time could be tedious and error prone. Let's see how Istio can improve your experience doing upgrades.

Reviewer 2: ```suggestion
To support this revision-based upgrade, Istio introduced a `istio.io/rev` label for namespaces. This indicates which control plane the sidecar proxies in that namespace should connect to. For example, a label of `istio.io/rev=1-9-3` indicates the control plane revision `1-9-3` should inject the data plane for workloads in that namespace.

If you wanted to upgrade the data-plane proxies for a particular namespace, you would update the `istio.io/rev` label to point to a new version, such as `istio.io/rev=1-10-0`. Manually changing (or even trying to orchestrate) changes of labels across a large number of namespaces can be error-prone and lead to unintended downtime.
```

I'm not sure about the first sentence.  It talks first about _what control plane the proxy should connect to_, but then switches to _what version should be injected_.  These two should probably be the same, but it's not clear in this text that they are.

Also moved a sentence up from below.

## Introducing Revision Tags

> [In Istio 1.10](/news/releases/1.10.x/announcing-1.10/), we've improved the revision feature [with a new feature called _revision tags_](/docs/setup/upgrade/canary/#stable-revision-labels-experimental). A revision `tag` reduces the number of changes an operator has to make to use revisions and safely upgrade an Istio control plane. With a `tag`, you can point to a revision and use the `tag` as the label for your namespaces. That way you don't have to change the labels on a namespace and minimize the number of manual steps and configuration changes upgrading. Manually changing (or even trying to orchestrate) changes of labels across a large number of namespaces can be error-prone and lead to unintended downtime.

Reviewer 2: ```suggestion
[In Istio 1.10](/news/releases/1.10.x/announcing-1.10/), we've improved revision-based upgrades with a new feature called _[revision tags](/docs/setup/upgrade/canary/#stable-revision-labels-experimental)_. A revision tag reduces the number of changes an operator has to make to use revisions, and safely upgrade an Istio control plane. You use the tag as the label for your namespaces, and assign a revision to that tag. This means you don't have to change the labels on a namespace while upgrading, and minimizes the number of manual steps and configuration changes. 
```

> For example, you can define a `tag` named `prod-stable` and point it to the `1-9-3` revision of a control plane. You can also define another `tag` named `prod-canary` and also point it to a new `1-10-0` revision. You may have a lot of important namespaces in your cluster and can label those namespaces with `istio.io/rev=prod-stable`. You may have other namespaces where you are willing to canary control plane changes and can label that `istio.io/rev=prod-canary`. In both of these scenarios, the tag will indirectly associate those namespaces with the `1-9-3` revision for `prod-stable` and `1-10-0` for `prod-canary` respectively.
> 
> {{< image link="./tags.png" caption="Stable revision tags" >}}
> 
> Once you've determined the new control plane is suitable for the rest of the `prod-stable` namespaces, you can change the tag to point to the new revision. This will enable you to update the namespaces labeled `prod-stable` to the new `1-10-0` revision without making any changes to the labels on the namespace.

Reviewer 2: ```suggestion
For example, you can define a tag named `prod-stable` and point it to the `1-9-3` revision of a control plane. You can also define another tag named `prod-canary` which points to the `1-10-0` revision. You may have a lot of important namespaces in your cluster, and you can label those namespaces with `istio.io/rev=prod-stable`. In other namespaces you may be willing to test the new version of Istio, and you can label that namespace `istio.io/rev=prod-canary`. The tag will indirectly associate those namespaces with the `1-9-3` revision for `prod-stable` and `1-10-0` for `prod-canary` respectively.

{{< image link="./tags.png" caption="Stable revision tags" >}}

Once you've determined the new control plane is suitable for the rest of the `prod-stable` namespaces, you can change the tag to point to the new revision. This enables you to update all the namespaces labeled `prod-stable` to the new `1-10-0` revision without making any changes to the labels on the namespaces.
```

{{< image link="./tags-updated.png" caption="Updated revision tags" >}}

## Stable revision tags in action

> For example, to create a new `prod-stable` tag for a revision `1-9-3`, you would run the following command:

Reviewer 2: ```suggestion
To create a new `prod-stable` tag for a revision `1-9-3`, run the following command:
```

{{< text bash >}}
$ istioctl x revision tag set prod-stable --revision 1-9-3
{{< /text >}}

You can then label your namespaces with the `istio.io/rev=prod-stable` label. Note, if you installed a `default` revision (i.e., no revision) of Istio, you will first have to remove the standard injection label:

{{< text bash >}}
$ kubectl label ns istioinaction istio-injection-
$ kubectl label ns istioinaction istio.io/rev=prod-stable
{{< /text >}}

You can list the tags in your mesh with the following:

{{< text bash >}}
$ istioctl x revision tag list

TAG         REVISION NAMESPACES
prod-stable 1-9-3
{{< /text >}}

A tag is implemented with a `MutatingWebhookConfiguration`. You can verify a corresponding `MutatingWebhookConfiguration` has been created:

{{< text bash >}}
$ kubectl get MutatingWebhookConfiguration

NAME                             WEBHOOKS   AGE
istio-revision-tag-prod-stable   2          75s
istio-sidecar-injector           1          5m32s
{{< /text >}}

Let's say you are trying to canary a new revision of the control plane based on 1.10.0. First you would install the new version using a revision:

{{< text bash >}}

> $  istioctl install -y --set profile=minimal --revision 1-10-0

Reviewer 2: ```suggestion
$ istioctl install -y --set profile=minimal --revision 1-10-0
```

{{< /text >}}

You can create a new tag called `prod-canary` and point that to your `1-10-0` revision:

{{< text bash >}}

> $  istioctl x revision tag set prod-canary --revision 1-10-0

Reviewer 2: ```suggestion
$ istioctl x revision tag set prod-canary --revision 1-10-0
```

{{< /text >}}

Then label your namespaces accordingly:

{{< text bash >}}
$ kubectl label ns istioinaction-canary istio.io/rev=prod-canary
{{< /text >}}

> If you list out the tags in our system, you should see two stable tags pointing to different revisions:

Reviewer 2: ```suggestion
If you list out the tags in our mesh, you will now see two tags pointing to two different revisions:
```

Reviewer: s/our mesh/your mesh

{{< text bash >}}
$ istioctl x revision tag list

TAG         REVISION NAMESPACES
prod-stable 1-9-3
prod-canary 1-10-0
{{< /text >}}

Any of the namespaces that you have labeled with `istio.io/rev=prod-canary` will be injected by the control plane that corresponds to the `prod-canary` stable tag name (which in this example points to the `1-10-0` revision). When you're ready, you can switch the `prod-stable` tag to the new control plane with:

{{< text bash >}}

> $  istioctl x revision tag set prod-stable --revision 1-10-0 --overwrite

Reviewer 2: ```suggestion
$ istioctl x revision tag set prod-stable --revision 1-10-0 --overwrite
```

{{< /text >}}

> When both the `prod-stable` and `prod-canary` tags both point to the same new revision, it may be safe to remove the old revision.

Reviewer 2: Missing from this story: what happens when you've made that change?  Do workloads update automatically, or do you have to do a rolling restart?

## Wrapping up

> Using revisions makes it safe to canary changes to an Istio control plane. In large environments with lots of namespaces, you may prefer to use stable tags as we've introduced in this blog to remove the number of moving pieces and simplify any automation you may build around updating an Istio control plane. [Check out the Istio 1.10 release](/news/releases/1.10.x/announcing-1.10/) [and the new `tag` feature](/docs/setup/upgrade/canary/#stable-revision-labels-experimental) and give us feedback!

Reviewer 2: ```suggestion
Using revisions makes it safer to canary changes to an Istio control plane. In large environments with lots of namespaces, you may prefer to use stable tags, as we've introduced in this blog, to remove the number of moving pieces and simplify any automation you may build around updating an Istio control plane. Please check out the [1.10 release](/news/releases/1.10.x/announcing-1.10/) [and the new tag feature](/docs/setup/upgrade/canary/#stable-revision-labels-experimental) and give us your feedback!
```

---
title: "Istio control plane canary with revisions and tags"
description: Learn how to take more control over your Istio control-plane upgrades.
publishdate: 2021-05-26
attribution: "Christian Posta (Solo.io), Lin Sun (Solo.io), Sam Naser (Google)"
keywords: [upgrades,revisions,operations,canary]
---

> Like all security software, your service mesh should be kept up-to-date. The Istio community [releases new versions every quarter](/docs/releases/supported-releases/), with regular patch releases for bug fixes [and security vulnerabilities](/docs/releases/security-vulnerabilities/). The operator of a service mesh will need to upgrade the control plane and data plane components many times. You must take care when upgrading, as a mistake could affect your business traffic. Istio has many mechanisms to make it safe to perform upgrades in a controlled manner, and in Istio 1.10 we further improve this operational experience.

Reviewer: Does it make sense to change the security vulnerability link to point to https://istio.io/latest/blog/2021/patch-tuesdays/ `Updates to how Istio security releases are handled: Patch Tuesday, embargoes, and 0-days`?  I think either link is probably good, but the blog post does mention Patch Tuesday scheduling.

## Background

> [In Istio 1.6](/news/releases/1.6.x/announcing-1.6/change-notes/), we added [basic support for upgrading the service mesh following a canary pattern, using revisions](/blog/2020/multiple-control-planes/). Using this approach, you can run multiple control planes side-by-side without impacting an existing deployment and slowly migrate workloads from the old control plane to the new.

Reviewer: I'm thinking there shouldn't be a `,` after pattern.

If you wanted to upgrade the data-plane proxies for a particular namespace, you would update the `istio.io/rev` label to point to a new version, such as `istio.io/rev=1-10-0`. Manually changing (or even trying to orchestrate) changes of labels across a large number of namespaces can be error-prone and lead to unintended downtime.

## Introducing Revision Tags

[In Istio 1.10](/news/releases/1.10.x/announcing-1.10/), we've improved revision-based upgrades with a new feature called _[revision tags](/docs/setup/upgrade/canary/#stable-revision-labels-experimental)_. A revision tag reduces the number of changes an operator has to make to use revisions, and safely upgrade an Istio control plane. You use the tag as the label for your namespaces, and assign a revision to that tag. This means you don't have to change the labels on a namespace while upgrading, and minimizes the number of manual steps and configuration changes.

> For example, you can define a tag named `prod-stable` and point it to the `1-9-5` revision of a control plane. You can also define another tag named `prod-canary` which points to the `1-10-0` revision. You may have a lot of important namespaces in your cluster, and you can label those namespaces with `istio.io/rev=prod-stable`. In other namespaces you may be willing to test the new version of Istio, and you can label that namespace `istio.io/rev=prod-canary`. The tag will indirectly associate those namespaces with the `1-9-5` revision for `prod-stable` and `1-10-0` for `prod-canary` respectively.

Reviewer 3: I'm not sure 'prod-stable' is the best name. 

I think simpler 'stable' and 'canary' will be easier to understand and fit most use cases. And if a user needs custom names - they can still use them.

> {{< image link="./tags.png" caption="Stable revision tags" >}}

Reviewer: The picture shows 1.9.3, but the words all mention 1.9.5. People will figure out what is meant, but might be better to have them match,.

Reviewer: The problem with the new updated picture is everything is pointing to the 1.10.0 istiod.

{{< image link="./tags-updated.png" caption="Updated revision tags" >}}

Once you're satisfied with the upgrade to the new control-plane revision, you can remove the old control plane.

## Stable revision tags in action

To create a new `prod-stable` tag for a revision `1-9-5`, run the following command:

{{< text bash >}}
$ istioctl x revision tag set prod-stable --revision 1-9-5
{{< /text >}}

You can then label your namespaces with the `istio.io/rev=prod-stable` label. Note, if you installed a `default` revision (i.e., no revision) of Istio, you will first have to remove the standard injection label:

{{< text bash >}}
$ kubectl label ns istioinaction istio-injection-
$ kubectl label ns istioinaction istio.io/rev=prod-stable
{{< /text >}}

You can list the tags in your mesh with the following:

{{< text bash >}}
$ istioctl x revision tag list

TAG         REVISION NAMESPACES
prod-stable 1-9-5

> {{< /text >}}

Reviewer: Is this one missing the `istioinaction` namespace?

{{< text bash >}}
$ istioctl install -y --set profile=minimal --revision 1-10-0
{{< /text >}}

You can create a new tag called `prod-canary` and point that to your `1-10-0` revision:

{{< text bash >}}
$ istioctl x revision tag set prod-canary --revision 1-10-0
{{< /text >}}

Then label your namespaces accordingly:

{{< text bash >}}
$ kubectl label ns istioinaction-canary istio.io/rev=prod-canary
{{< /text >}}

If you list out the tags in your mesh, you will see two stable tags pointing to two different revisions:

{{< text bash >}}
$ istioctl x revision tag list

TAG         REVISION NAMESPACES
prod-stable 1-9-5    istioinaction
prod-canary 1-10-0   istioinaction-canary
{{< /text >}}

Any of the namespaces that you have labeled with `istio.io/rev=prod-canary` will be injected by the control plane that corresponds to the `prod-canary` stable tag name (which in this example points to the `1-10-0` revision). When you're ready, you can switch the `prod-stable` tag to the new control plane with:

{{< text bash >}}
$ istioctl x revision tag set prod-stable --revision 1-10-0 --overwrite
{{< /text >}}

> Any time you switch a tag to point to a new revision, you should restart the workloads in any respective namespace to pick up the new revision's proxy.

Reviewer: Here you say should restart, while above you say will need to restart. I believe it is the later, so change should to will need.

Reviewer 3: How about an example - using rolling upgrade ?

> When both the `prod-stable` and `prod-canary` no longer point to the old revision, it may be safe to remove the old revision like the following:

Reviewer 4: ```suggestion
When both the `prod-stable` and `prod-canary` no longer point to the old revision, it may be safe to remove the old revision as follows:
```

> When both the `prod-stable` and `prod-canary` tags both point to the same new revision, it may be safe to remove the old revision like the following:

Reviewer: Would it be better to say When both the `prod-stable` and `prod-canary` tags no longer point to the old version, ...  Or maybe When no revisions or revision tags point to the old version,... ?

Reviewer 3: Yes, user may have many other tags - "when no tags point to a revision" is better.

May also note somewhere that 'Users should make sure the revision is not directly used by a namespace 
or object' ( since in 1.10 you can also set it as a label on a workload, without any label in the namespace ).

keywords: [upgrades,revisions,operations,canary]
---

The Istio community [releases new versions of Istio](/about/supported-releases/) every quarter and sometimes between the normal release cadence for bug fixes [or CVE patches](/about/security-vulnerabilities/). The operator of an Istio service mesh needs to upgrade the control plane and data plane components during the life of the service mesh. You must take care when upgrading a service mesh as a mistake could take down the entire mesh and affect your business traffic. Istio has a couple mechanisms to make it safe to perform upgrades in a controlled manner, and in Istio 1.10 we further improve this operational experience.

## Background

In Istio 1.6, [we added basic support for upgrading the service mesh](/news/releases/1.6.x/announcing-1.6/change-notes/) following a [canary pattern using revisions](/blog/2020/multiple-control-planes/). Using this approach, you can run multiple control planes side-by-side without impacting an existing deployment and slowly migrate workloads from an old control plane to the new control plane.

> To support this revision-based upgrade, Istio introduced the `istio.io/rev` label for namespaces which indicates to which control plane the workload data-plane proxies in that namespace should connect. For example, a label of `istio.io/rev=1-9-3` indicates the control plane revision `1-9-3` should inject the data plane for workloads in that namespace.

Reviewer 4: ```suggestion
To support this revision-based upgrade, Istio introduced the `istio.io/rev` label for namespaces which determines which control plane the sidecar proxies in that namespace should connect to. For example, labeling a namespace `istio.io/rev=1-9-3` indicates the control plane revision `1-9-3` should inject the sidecar proxies for workloads in that namespace.
```

> If you wanted to upgrade the data-plane proxies for a particular namespace, you would update the `istio.io/rev` label to point to a new version such as `istio.io/rev=1-10-0`. Across many namespaces, updating labels one at a time could be tedious and error prone. Let's see how Istio can improve your experience doing upgrades.

Reviewer 4: ```suggestion
If you wanted to upgrade the data-plane proxies for a particular namespace, you would update the `istio.io/rev` label to point to a new version such as `istio.io/rev=1-10-0`. Across many namespaces, updating labels one at a time could be tedious and error prone.
```

## Introducing Revision Tags

[In Istio 1.10](/news/releases/1.10.x/announcing-1.10/), we've improved the revision feature [with a new feature called _revision tags_](/docs/setup/upgrade/canary/#stable-revision-labels-experimental). A revision `tag` reduces the number of changes an operator has to make to use revisions and safely upgrade an Istio control plane. With a `tag`, you can point to a revision and use the `tag` as the label for your namespaces. That way you don't have to change the labels on a namespace and minimize the number of manual steps and configuration changes upgrading. Manually changing (or even trying to orchestrate) changes of labels across a large number of namespaces can be error-prone and lead to unintended downtime.

> For example, you can define a `tag` named `prod-stable` and point it to the `1-9-3` revision of a control plane. You can also define another `tag` named `prod-canary` and also point it to a new `1-10-0` revision. You may have a lot of important namespaces in your cluster and can label those namespaces with `istio.io/rev=prod-stable`. You may have other namespaces where you are willing to canary control plane changes and can label that `istio.io/rev=prod-canary`. In both of these scenarios, the tag will indirectly associate those namespaces with the `1-9-3` revision for `prod-stable` and `1-10-0` for `prod-canary` respectively.

Reviewer 4: Sorry, again not sure why I can't comment directly on the "revision tags in action" section but when you do `istioctl x revision tag list` after labeling a namespace with the `prod-stable` tag, the output should show that namespace next to the tag under the `NAMESPACES` column but doesn't seem to?

Also, at the end of that section when saying it's safe to remove the old revision, maybe document how to do that? (`istioctl x uninstall --revision 1-9-5` or whatever)
